rfcs/sdk-spec.md

```diff
@@ -0,0 +1,1043 @@
+* Job (P2) - a long-running compute workload that can be run on demand
+* Workflow (P2) - a task orchestration engine, similar to AWS Step Functions, Azure Logic Apps, GCP Workflows
+* Secret (P2) - a secret value, similar to AWS Secrets Manager, Azure Key Vault, GCP Secret Manager
+* Stream (P2) - a stream of events, similar to AWS Kinesis, Azure Event Hubs, GCP Pub/Sub and Dataflow
+* OnDeploy (P2) - a variation of Function that runs every time the app is deployed
+* GraphQLApi (P2) - a GraphQL API, similar to AWS AppSync
+
+### Resources planned as third party libraries
+
+* DynamoDBTable
+* Redis
+* MongoDB
+* GithubRepo
+* Authorization/authentication related resources
+
+## Bucket
+
+The bucket resource represents an object store that can be used to store and retrieve arbitrary data.
+A bucket can be used to store text files, images, videos, and any other type of data.
+You can think of a bucket like hash map in the cloud whose data is typically distributed across multiple machines for high availability.
+
+<!--
+All code snippets should be in Wing - it just says "ts" for syntax highlighting
+-->
+
+```ts
+struct BucketProps {
+  /**
+   * Whether the bucket is public.
+   * @default false
+   */
+  public: bool?;
+}
+
+struct BucketOnUploadProps { /* elided */ }
+struct BucketOnDeleteProps { /* elided */ }
+struct BucketOnUpdateProps { /* elided */ }
+struct BucketOnEventProps { /* elided */ }
+
+resource Bucket {
+  init(props: BucketProps = {});
+
+  /**
+   * Run an inflight whenever a file is uploaded to the bucket.
+   */
+  on_upload(fn: inflight (key: str) => void, opts: BucketOnUploadProps?): void;
+
+  /**
+   * Run an inflight whenever a file is deleted from the bucket.
+   */
+  on_delete(fn: inflight (key: str) => void, opts: BucketOnDeleteProps?): void;
+
+  /**
+   * Run an inflight whenever a file is updated in the bucket.
+   */
+  on_update(fn: inflight (key: str) => void, opts: BucketOnUpdateProps?): void;
+
+  /**
+   * Run an inflight whenever a file is created, uploaded, or deleted from the bucket.
+   */
+  on_event(fn: inflight (event: BucketEvent) => void, opts: BucketOnEventProps?): void;
+
+  /**
+   * Upload a file to the bucket.
+   */
+  inflight put(key: str, value: Serializable): void;
+
+  /**
+   * Get a file from the bucket.
+   */
+  inflight get(key: str): Serializable;
+
+  /**
+   * Delete a file from the bucket.
+   */
+  inflight delete(key: str): void;
+
+  /**
+   * List all files in the bucket with the given prefix.
+   */
+  inflight list(prefix: str?): Iterator<str>;
+
+  /**
+   * Returns a url to the given file.
+   * @throws Will throw if the public is not public.
+   */
+  inflight public_url(key: str): str;
+
+  /**
+   * Returns a presigned url to the given file. This URL can be used to access
+   * the file by anyone until it expires (defaults to 24 hours).
+   */
+  inflight presigned_url(key: str, duration?: duration): str;
```

Comment:
does calling this inflight `presigned_url` go against cloud/target agnostic? 

To me the naming makes sense but that might just be my AWS bias showing 😅 the same concept in GCP is just called `signed url` and in Azure (who just has to be different) calls this `shared access signature`

Reply:
Good callout - it didn't occur to me that presigned url could have been an AWS-ism. On Google, I'm seeing more search results for "signed url" than "presigned url" or "pre-signed url", so I'll change the name 😛